### PR TITLE
[Core: User] Add type declarations to function signatures

### PR DESCRIPTION
**Pull request by johnsaigle**
### Brief summary of changes
Types are good!

Further PRs to follow if Travis fails.

### Related 
- [ ] #3878, #4020 

##### Bugs Found
* #4033

### To test this change...

- [ ] Everything should stay the same.

### CODE REVIEW

**Comment by driusan** [comment]
Should this declare strict_types for the file?

File: php/libraries/User.class.inc [hunk 1]
```diff
@@ -182,13 +182,13 @@ class User extends UserPermissions
...
      *
-     * @param string $var Name of variable to get
+     * @param string|null $var Name of variable to get
      *
      * @note   Call without any parameters to get the entire user data array
      * @return mixed
      * @access public
      */
-    function getData($var = null)
+    function getData(?string $var = null)
```
Comment (kongtiaowang) [inline php/libraries/User.class.inc]: Why this function doesn't have a return type?

Comment (PapillonMcGill) [reply php/libraries/User.class.inc]: @kongtiaowang 
Because it is *mixed* and the declaration do not support mixed return type (yet).

**Comment by driusan** [comment]
@johnsaigle conflicts, needs rebase

File: php/libraries/User.class.inc (hunk 1 above)
Comment (HenriRabalais) [reply php/libraries/User.class.inc]: Do you need the `?` for type hinting function parameters? I was under the impression that was only for return type declarations.

Comment (johnsaigle) [reply php/libraries/User.class.inc]: @HenriRabalais I'm not actually sure about that... I don't know whether the Factory should accept null as a parameter even though the PHPDOC says that it should

Comment (driusan) [reply php/libraries/User.class.inc]: `?type` means "null or type" while `type` means "must be type and only type"

**Comment by johnsaigle** [comment]
This will be rebased once #4033 is merged.

**Comment by johnsaigle** [comment]
@PapillonMcGill LORIS won't work if the parameter to the factory function isn't nullable. I'm not sure why but that's why this line was changed in the latest commit. See discussion here: https://github.com/aces/Loris/pull/3878#discussion_r211684049

**Comment by HenriRabalais** [comment]
@johnsaigle I don't the suggestion is the make the factory function non-nullable, but rather have it default to null if no value is passed:
i.e. `public static function &factory(?string $username=null): \User`

**Comment by PapillonMcGill** [comment]
@johnsaigle I'm not asking to make them `not nullable` but to do like line 131 `?string $username = null`
just like Henry wrote in the previous comment.

File: php/libraries/User.class.inc [hunk 2]
```diff
@@ -29,24 +29,24 @@ class User extends UserPermissions
      * @var    array
      * @access private
      */
-    var $userInfo = array();
+    private $_userInfo = array();
```
Comment (driusan) [inline php/libraries/User.class.inc]: protected

File: php/libraries/User.class.inc [hunk 3]
```diff
@@ -260,12 +262,12 @@ class User extends UserPermissions
      *
      * @return string
      */
-    function getSiteNames()
+    function getSiteNames(): string
```
Comment (driusan) [inline php/libraries/User.class.inc]: getSiteName*s* is plural, shouldn't this be an array of strings, not a string?

Comment (johnsaigle) [reply php/libraries/User.class.inc]: #4118 Addresses this change and fixes the calling code.

**Comment by driusan** [comment]
@johnsaigle conflicts

**Pull request merged**